File: src/sdk/typescript/src/discovery.ts
import { accessSync, constants, existsSync } from "node:fs";
import { delimiter, dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { BridgeBinaryNotFoundError, BridgeBinaryPermissionError } from "./errors.js";

function platformKey(): string {
  const os = process.platform === "win32" ? "windows" : process.platform === "darwin" ? "darwin" : process.platform === "linux" ? "linux" : undefined;
  const arch = process.arch === "x64" ? "amd64" : process.arch === "arm64" ? "arm64" : undefined;
  if (!os || !arch) throw new BridgeBinaryNotFoundError(`unsupported platform: ${process.platform}/${process.arch}`);
  return `${os}-${arch}`;
}
function pathCandidate(): string | undefined {
  const exe = process.platform === "win32" ? "bridge.exe" : "bridge";
  for (const dir of (process.env.PATH ?? "").split(delimiter)) {
    const p = join(dir, exe); if (existsSync(p)) return p;
  }
  return undefined;
}
export function resolveBinary(binaryPath?: string): string {
  const exe = process.platform === "win32" ? "bridge.exe" : "bridge";
  const moduleDir = dirname(fileURLToPath(import.meta.url));
  const bundled = join(moduleDir, "..", "bin", platformKey(), exe);
  const candidates = [binaryPath, process.env.BRIDGE_BINARY, bundled, pathCandidate()].filter((v): v is string => Boolean(v));
  for (const p of candidates) {
    if (!existsSync(p)) continue;
    if (process.platform !== "win32") {
      try { accessSync(p, constants.X_OK); } catch { throw new BridgeBinaryPermissionError(`BRIDGE binary is not executable: ${p}`); }
    }
    return p;
  }
  throw new BridgeBinaryNotFoundError("BRIDGE binary was not found");
}

File: src/sdk/python/src/bridge_sdk/discovery.py
from __future__ import annotations
import os, platform, shutil
from importlib.resources import files
from pathlib import Path
from .errors import BridgeBinaryNotFoundError, BridgeBinaryPermissionError

def _platform_key() -> str:
    system = platform.system().lower()
    machine = platform.machine().lower()
    os_key = {"linux":"linux", "windows":"windows", "darwin":"darwin"}.get(system)
    arch_key = {"x86_64":"amd64", "amd64":"amd64", "aarch64":"arm64", "arm64":"arm64"}.get(machine)
    if not os_key or not arch_key:
        raise BridgeBinaryNotFoundError(f"unsupported platform: {system}/{machine}")
    return f"{os_key}-{arch_key}"

def resolve_binary(binary_path: str | os.PathLike[str] | None = None) -> Path:
    candidates = []
    if binary_path:
        candidates.append(Path(binary_path))
    if os.environ.get("BRIDGE_BINARY"):
        candidates.append(Path(os.environ["BRIDGE_BINARY"]))
    exe = "bridge.exe" if os.name == "nt" else "bridge"
    candidates.append(Path(str(files("bridge_sdk").joinpath("bin", _platform_key(), exe))))
    path_hit = shutil.which("bridge")
    if path_hit:
        candidates.append(Path(path_hit))
    for candidate in candidates:
        if candidate.is_file():
            if os.name != "nt" and not os.access(candidate, os.X_OK):
                raise BridgeBinaryPermissionError(f"BRIDGE binary is not executable: {candidate}")
            return candidate
    raise BridgeBinaryNotFoundError("BRIDGE binary was not found")
